js/packages/openinference-vercel/src/utils.ts

```diff
@@ -0,0 +1,439 @@
+      acc[`${MESSAGE_PREFIX}.${SemanticConventions.MESSAGE_CONTENT}`] =
+        message.content;
+    }
+    acc[
+      `${SemanticConventions.LLM_INPUT_MESSAGES}.${index}.${SemanticConventions.MESSAGE_ROLE}`
+    ] = typeof message.role === "string" ? message.role : undefined;
+    return acc;
+  }, {});
+};
+
+/**
+ * Gets the output_messages tool_call OpenInference attributes
+ * @param toolCalls the attribute value of the Vercel result.toolCalls
+ * @returns output_messages tool_call attributes
+ */
+const getToolCallMessageAttributes = (toolCalls?: AttributeValue) => {
+  if (typeof toolCalls !== "string") {
+    return null;
+  }
+
+  const parsedToolCalls = JSON.parse(toolCalls);
+
+  if (!isArrayOfObjects(parsedToolCalls)) {
+    return null;
+  }
+
+  const OUTPUT_MESSAGE_PREFIX = `${SemanticConventions.LLM_OUTPUT_MESSAGES}.0`;
+  return {
+    [`${OUTPUT_MESSAGE_PREFIX}.${SemanticConventions.MESSAGE_ROLE}`]:
+      "assistant",
+    ...parsedToolCalls.reduce((acc: Attributes, toolCall, index) => {
+      const TOOL_CALL_PREFIX = `${OUTPUT_MESSAGE_PREFIX}.${SemanticConventions.MESSAGE_TOOL_CALLS}.${index}`;
+      return {
+        ...acc,
+        [`${TOOL_CALL_PREFIX}.${SemanticConventions.TOOL_CALL_FUNCTION_NAME}`]:
+          isAttributeValue(toolCall.toolName) ? toolCall.toolName : undefined,
+        [`${TOOL_CALL_PREFIX}.${SemanticConventions.TOOL_CALL_FUNCTION_ARGUMENTS_JSON}`]:
+          toolCall.args != null ? JSON.stringify(toolCall.args) : undefined,
+      };
+    }, {}),
+  };
+};
+
+/**
+ * Gets the OpenInference metadata attributes
+ * Both vercel and OpenInference attach metadata attributes to spans in a flat structure
+ * @example Vercel: ai.telemetry.metadata.<metadataKey>
+ * @example OpenInference: metadata.<metadataKey>
+ * @param attributes the initial attributes of the span
+ * @returns the OpenInference metadata attributes
+ */
+const getMetadataAttributes = (attributes: Attributes) => {
+  const metadataAttributeKeys = Object.keys(attributes)
+    .filter((key) => key.startsWith(VercelSemanticConventions.METADATA))
+    .map((key) => ({ key: key.split(".")[2], value: attributes[key] }));
+  if (metadataAttributeKeys.length === 0) {
+    return null;
+  }
+  return metadataAttributeKeys.reduce(
+    (acc, { key, value }) => ({
+      ...acc,
+      [`${SemanticConventions.METADATA}.${key}`]: value,
+    }),
+    {},
+  );
+};
+
+/**
+ * Gets the OpenInference attributes associated with the span from the initial attributes
+ * @param attributesWithSpanKind the initial attributes of the span and the OpenInference span kind
+ * @param attributesWithSpanKind.initialAttributes the initial attributes of the span
+ * @param attributesWithSpanKind.spanKind the OpenInference span kind
+ * @returns The OpenInference attributes associated with the span
+ */
+export const getOpenInferenceAttributes = ({
+  initialAttributes,
+  spanKind,
+}: {
+  initialAttributes: Attributes;
+  spanKind?: OpenInferenceSpanKind;
```

Comment:
I guess I find this a tad confusing having this as a parameter here. Is there not a way to just create a single transform function rather than having a two step process?

Reply:
yeah didn't feel great about this one either, can just move it in and compute it up front - it's only used for one thing to make the switch a little less crazy but not super important